lpcprog.1
.TH LPCPROG "1" "July 2014" "LPC Tools 1.07" "User development Tools"
.SH NAME
lpcprog \- NXP's LPC micro-controllers flasher
.SH SYNOPSIS
.B lpcprog
\fI\-d serial_device\fR \fI\-c command\fR [\fIOPTIONS\fR] ... [\fIFILE\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Use ISP protocol to flash LPC micro-controllers over serial line.
.PP
For all commands the target micro-controller must be in ISP mode and waiting for
synchronisation ISP command (default state when entering ISP mode). See ISP MODE
AND CODE READ PROTECTION section for more details.
.PP
A FILE argument \fBmust\fR be given for dump and flash commands. For flash command
this file is the data (program) to be written to LPC internal Flash memory. For dump
command, it will store the actual content of the LPC internal flash memory.
.PP
Mandatory arguments to long options are mandatory for short options too.
.TP
\fB\-p\fR, \fB\-\-parts\fR==\fIPARTS\fR
Use PARTS as parts desription file instead of the default ones. See PARTS DESCRIPTION
FILES section for more information on default parts desription files and order of
precedence.
.TP
\fB\-d\fR, \fB\-\-device\fR=\fIDEV\fR
Use DEV as host serial line to program the target. Full path must be provided.
Device node files are usually located in /dev/ directory.
.TP
\fB\-c\fR, \fB\-\-command\fR=\fICOMMAND\fR
Command to execute. COMMAND must be one of \fBid\fR, \fBdump\fR, \fBflash\fR,
\fBblank\fR or \fBgo\fR.
See COMMANDS section for commands description.
.TP
\fB\-b\fR, \fB\-\-baudrate\fR=\fIBAUD\fR
Use BAUD as the baudrate for communication with the target device. Defaults to
115200.
.TP
\fB\-t\fR, \fB\-\-trace\fR
Turn on trace output of serial communication with target device
.TP
\fB\-f\fR, \fB\-\-freq\fR=\fIFREQ\fR
Use FREQ (KHz) as the oscilator frequency of target device. Defaults to 10000 KHz
.TP
\fB\-n\fR, \fB\-\-no\-user\-code\fR
Do not compute a valid user code for exception vector 7. See USER CODE section.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display help information and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
Display version information and exit
.SH COMMANDS
.PP
The command must be one of \fBid\fR, \fBdump\fR, \fBflash\fR, \fBblank\fR or \fBgo\fR.
.TP
\fBid\fR
The \fBid\fR command displays the LPC part identification ID, the uid (unique ID),
and the bootloader version of the connected target device.
.TP
\fBdump\fR
Dump the whole connected target's flash memory content to the given file. This command
requires a file argument.
.TP
\fBflash\fR
Flash the content of the file given as argument to the beginning of the connected
target's flash memory. Automatic computation of the User Code is made and User Code is
stored in the 7th exception vector. Use \fB\-n\fR option to prevent User Code
modification. If you need to write your file to a different flash section, use the
\fBlpcisp\fR tool.
.TP
\fBblank\fR
Erase the whole flash.
.TP
\fBgo\fR
Unsupported Yet. Reset the target using hardware reset button or power cycle the
device to start the program.
.SH "PARTS DESCRIPTION FILES"
Default parts description files are /etc/lpctools_parts.def or ./lpctools_parts.def
The parts description file is parsed for LPC device description for dump, blank, and
flash commands. If none of the defaults exist and no existing file is supplied
using \fB\-p\fR option, these command cannot execute. If the part description is not
found in the first found file, the command will not execute either.
If no \fB\-p\fR option is provided, the program looks for lpctools_parts.def in the
current directory first, and then in the /etc/ directory.
.PP
The content of parts description files is a list of LPC parts description.
Each line holds one part description. Lines starting by # are comments.
Refer to the default file found in the /etc/ directory for more information on the
line format.
.SH "USER CODE"
A valid User Code is requiered by the LPC bootloader to start the internal user
program. User Code is considered valid if the reserved exception vector location 7
(offset 0x 0000 001C in the vector table) contains the 2’s complement of the
check-sum of table entries 0 through 6. This causes the checksum of the first 8
table entries to be 0.
.PP
If the signature is not valid the LPC enters ISP mode.
.PP
The default behavior of \fBlpcprog\fR is to compute a valid code for the 7th
exception vector and modify the programed binary data to include this valid code.
You can prevent this by using \fB\-n\fR option.
.SH "ISP MODE AND CODE READ PROTECTION"
Code Read Protection is a mechanism that allows "the user" to enable different
levels of "security" in the system so that access to the on-chip flash and use of
the ISP can be restricted. When needed, CRP is invoked by programing a specific
pattern in flash location at 0x0000 02FC. IAP commands are not affected by the
code read protection.
.PP
If your device Does not enter ISP state, or if some commands do not succeed, it may
be because the CRP mechanism is in use on your device.
.PP
Code Read Protection will not be implemented as it goes against the license of this
tool and against the free software philosophy.
.SH AUTHOR
Written by Nathael Pajani.
.SH "REPORTING BUGS"
Report bugs to bug@techno-innov.fr
.br
Techno-Innov home page: <http://www.techno-innov.fr/>
.br
Software Git Repository: <http://git.techno-innov.fr/>
.SH COPYRIGHT
Copyright \(co 2013 Techno-Innov
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
lpcisp(1)


lpcprog.c
/*********************************************************************
 *
 *   ISP (In-System Programming) tool for NXP LPC Microcontrollers
 *
 * This tool does not give access to each isp command, instead it
 * provides wrappers for flashing a device.
 *
 *
 *  Copyright (C) 2012 Nathael Pajani <nathael.pajani@nathael.net>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *********************************************************************/


#include <stdlib.h> /* malloc, free */
#include <stdio.h>
#include <stdint.h>

#include <unistd.h> /* open, getopt, close, usleep */
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>

#include <errno.h>
#include <getopt.h>

#include <termios.h> /* for serial config */
#include <ctype.h>

#include <string.h> /* strncmp, strlen, strdup */

#include "isp_utils.h"
#include "isp_commands.h"
#include "prog_commands.h"
#include "parts.h"

#define PROG_NAME "LPC ISP Prog tool"
#define VERSION   "1.07"


void help(char *prog_name)
{
	fprintf(stderr, "---------------- "PROG_NAME" --------------------------------\n");
	fprintf(stderr, "Usage: %s -d <dev_name> -c <command> [options] [dump/prog file name]\n" \
		"  Default parts description files are /etc/lpctools_parts.def or ./lpctools_parts.def\n" \
		"  Default baudrate is B115200\n" \
		"  Default oscilator frequency used is 10000 KHz\n" \
		"  <command> is one of:\n" \
		"  \t dump, flash, id, blank, go\n" \
		"  \t dump file_name: dump flash content to 'file'\n" \
		"  \t flash file_name : put 'file' to flash, erasing requiered sectors\n" \
		"  \t blank : erase whole flash\n" \
		"  \t id : get all id information\n" \
		"  \t go : execute program from reset handler in thumb mode and open terminal\n" \
		"  Available options:\n" \
		"  \t -p | --parts=file : Parts description file (see defaults)\n" \
		"  \t -c | --command=cmd : \n" \
		"  \t -d | --device=dev_path : Host serial line used to program the device\n" \
		"  \t -b | --baudrate=N : Use this baudrate (Same baudrate must be used across whole session)\n" \
		"  \t -t | --trace : turn on trace output of serial communication\n" \
		"  \t -f | --freq=N : Oscilator frequency of target device\n" \
		"  \t -n | --no-user-code : do not compute a valid user code for exception vector 7\n" \
		"  \t -h | --help : display this help\n" \
		"  \t -v | --version : display version information\n", prog_name);
	fprintf(stderr, "-----------------------------------------------------------------------\n");
}

#define SERIAL_BAUD  B115200

int trace_on = 0;
int quiet = 0;
static int calc_user_code = 1; /* User code is computed by default */

char* parts_file_name = NULL;
#define DEFAULT_PART_FILE_NAME_ETC  "/etc/lpctools_parts.def"
#define DEFAULT_PART_FILE_NAME_CURRENT  "./lpctools_parts.def"

static int prog_connect_and_id(int freq);
static int prog_handle_command(char* cmd, int dev_id, int arg_count, char** args);

int main(int argc, char** argv)
{
	int baudrate = SERIAL_BAUD;
	int crystal_freq = 10000;
	char* isp_serial_device = NULL;
	int dev_id = 0;

	/* For "command" handling */
	char* command = NULL;
	char** cmd_args = NULL;
	int nb_cmd_args = 0;


	/* parameter parsing */
	while(1) {
		int option_index = 0;
		int c = 0;

		struct option long_options[] = {
			{"parts", required_argument, 0, 'p'},
			{"command", required_argument, 0, 'c'},
			{"device", required_argument, 0, 'd'},
			{"baudrate", required_argument, 0, 'b'},
			{"trace", no_argument, 0, 't'},
			{"freq", required_argument, 0, 'f'},
			{"no-user-code", no_argument, 0, 'n'},
			{"help", no_argument, 0, 'h'},
			{"version", no_argument, 0, 'v'},
			{0, 0, 0, 0}
		};

		c = getopt_long(argc, argv, "p:c:d:b:tf:nhv", long_options, &option_index);

		/* no more options to parse */
		if (c == -1) break;

		switch (c) {
			/* p, parts description file */
			case 'p':
				parts_file_name = strdup(optarg);
				break;

			/* c, command */
			case 'c':
				command = strdup(optarg);
				break;

			/* d, device */
			case 'd':
				isp_serial_device = strdup(optarg);
				break;

			/* b, baudrate */
			case 'b':
				baudrate = atoi(optarg);
				/* FIXME: validate baudrate */
				break;

			/* t, trace */
			case 't':
				trace_on = 1;
				break;

			/* f, freq */
			case 'f':
				crystal_freq = atoi(optarg);
				break;

			/* n, user-code */
			case 'n':
				calc_user_code = 0;
				break;

			/* v, version */
			case 'v':
				printf("%s Version %s\n", PROG_NAME, VERSION);
				return 0;
				break;

			/* h, help */
			case 'h':
			default:
				help(argv[0]);
				return 0;
		}
	}

	/* Command is mandatory */
	if (command == NULL) {
		printf("No command given. use -h or --help for help on available commands.\n");
		return -1;
	}
	/* Check for default parts file availability if none given as argument */
	if (parts_file_name == NULL) {
		FILE* parts_file = NULL;

		parts_file_name = DEFAULT_PART_FILE_NAME_CURRENT;
		parts_file = fopen(parts_file_name, "r");
		if (parts_file == NULL) {
			parts_file_name = DEFAULT_PART_FILE_NAME_ETC;
			parts_file = fopen(parts_file_name, "r");
		}
		if (parts_file == NULL) {
			printf("Unable to open default parts file to read LPC descriptions !\n");
			printf("Tried : '%s' and '%s'\n", DEFAULT_PART_FILE_NAME_ETC, DEFAULT_PART_FILE_NAME_CURRENT);
			return -1;
		}
		fclose(parts_file);
	}

	/* Open serial device */
	if (isp_serial_device == NULL) {
		printf("No serial device given, exiting\n");
		help(argv[0]);
		return 0;
	} else if (isp_serial_open(baudrate, isp_serial_device) != 0) {
		printf("Serial open failed, unable to initiate serial communication with target.\n");
		return -1;
	}

	if (trace_on) {
		printf("Serial device : %s\n", isp_serial_device);
		printf("Command : %s\n", command);
	}

	/* Parse remaining command line arguments (not options). They are command arguments */
	if (optind < argc) {
		nb_cmd_args = argc - optind;
		cmd_args = malloc(nb_cmd_args * sizeof(char *));
		if (trace_on) {
			printf("Command arguments :\n");
		}
		while (optind < argc) {
			static unsigned int idx = 0;
			cmd_args[idx++] = argv[optind++];
			if (trace_on) {
				printf("%s\n", cmd_args[idx - 1]);
			}
		}
	}

	/* First : sync with device */
	dev_id = prog_connect_and_id(crystal_freq);
	if (dev_id < 0) {
		printf("Unable to connect to target, consider hard reset of target or link\n");
		return -1;
	}

	if (command != NULL)  {
		int err = 0;
		err = prog_handle_command(command, dev_id, nb_cmd_args, cmd_args);
		if (err >= 0) {
			if (trace_on) {
				printf("Command \"%s\" handled OK.\n", command);
			}
		} else {
			printf("Error handling command \"%s\" : %d\n", command, err);
		}
	}


	if (cmd_args != NULL) {
		free(cmd_args);
	}
	isp_serial_close();
	return 0;
}

struct prog_command {
	int cmd_num;
	char* name;
};

static struct prog_command prog_cmds_list[] = {
	{0, "dump"},
	{1, "flash"},
	{2, "id"},
	{3, "blank"},
	{4, "go"},
	{5, NULL}
};

/*
 * Try to connect to the target and identify the device.
 * First try sync, and if it fails, try to read id twice. if both fail, then we are not connected
 */
static int prog_connect_and_id(int freq)
{
	int sync_ret = 0;

	/* Try to connect */
	sync_ret = isp_connect(freq, 1);
	/* Synchro failed or already synchronised ? */
	if (sync_ret < 0) {
		/* If already synchronized, then sync command and the next command fail.
		   Sync failed, maybe we are already sync'ed, then send one command for nothing */
		isp_cmd_part_id(1);
	}

	return isp_cmd_part_id(1);
}

static int prog_handle_command(char* cmd, int dev_id, int arg_count, char** args)
{
	int cmd_found = -1;
	int ret = 0;
	int index = 0;
	struct part_desc* part = NULL;

	if (cmd == NULL) {
		printf("prog_handle_command called with no command !\n");
		return -1;
	}

	part = find_part_in_file(dev_id, parts_file_name);
	if (part == NULL) {
		printf("Unknown part number : 0x%08x.\n", dev_id);
		return -2;
	}

	while ((cmd_found == -1) && (prog_cmds_list[index].name != NULL)) {
		if (strncmp(prog_cmds_list[index].name, cmd, strlen(prog_cmds_list[index].name)) == 0) {
			cmd_found = index;
			break;
		}
		index++;
	}
	if (cmd_found == -1) {
		printf("Unknown command \"%s\", use -h or --help for a list.\n", cmd);
		return -3;
	}

	switch (prog_cmds_list[cmd_found].cmd_num) {
		case 0: /* dump, need one arg : filename */
			if (arg_count != 1) {
				printf("command dump needs one arg (filename), got %d.\n", arg_count);
				return -4;
			}
			ret = dump_to_file(part, args[0]);
			break;

		case 1: /* flash, need one arg : filename */
			if (arg_count != 1) {
				printf("command flash needs one arg (filename), got %d.\n", arg_count);
				return -4;
			}
			ret = flash_target(part, args[0], calc_user_code);
			break;

		case 2: /* id : no args */
			ret = get_ids();
			break;

		case 3: /* blank : no args */
			ret = erase_flash(part);
			break;

		case 4: /* go : no args */
			ret = start_prog(part);
			break;
	}

	return ret;
}

